src/plugins/memdump/userhook.cpp

```diff
@@ -0,0 +1,671 @@
+#include <inttypes.h>
+#include <libvmi/libvmi.h>
+#include <libvmi/peparse.h>
+#include <assert.h>
+#include "memdump.h"
+
+static event_response_t usermode_hook_cb(drakvuf_t drakvuf, drakvuf_trap_info* info)
+{
+    hook_target_entry* target = (hook_target_entry*)info->trap->data;
+
+    if (target->pid != info->proc_data.pid)
+        return VMI_EVENT_RESPONSE_NONE;
+
+    PRINT_DEBUG("[MEMDUMP-USER] Usermode callback called %d!%s\n", info->proc_data.pid, info->trap->name);
+
+    vmi_instance_t vmi = drakvuf_lock_and_get_vmi(drakvuf);
+
+    bool is_syswow = false;
+    addr_t wow_ctx;
+
+    addr_t ethread = drakvuf_get_current_thread(drakvuf, info);
+
+    if (drakvuf_get_wow_context(drakvuf, ethread, &wow_ctx))
+    {
+        is_syswow = true;
+    }
+
+    access_context_t ctx =
+    {
+        .translate_mechanism = VMI_TM_PROCESS_DTB,
+        .dtb = info->regs->cr3,
+        .addr = info->regs->rsp
+    };
+
+    bool success = false;
+    addr_t ret_addr = 0;
+
+    if (is_syswow)
+    {
+        uint32_t ret_addr_tmp;
+
+        if (vmi_read_32(vmi, &ctx, &ret_addr_tmp) == VMI_SUCCESS)
+        {
+            success = true;
+            ret_addr = ret_addr_tmp;
+        }
+    }
+    else
+    {
+        success = vmi_read_64(vmi, &ctx, &ret_addr) == VMI_SUCCESS;
+    }
+
+    PRINT_DEBUG("[MEMDUMP] Return address %llx\n", (unsigned long long)ret_addr);
+    drakvuf_release_vmi(drakvuf);
+
+    mmvad_info_t mmvad;
+    if (!drakvuf_find_mmvad(drakvuf, info->proc_data.base_addr, ret_addr, &mmvad))
+    {
+        PRINT_DEBUG("[MEMDUMP] Failed to find MMVAD\n");
+        return VMI_EVENT_RESPONSE_NONE;
+    }
+
+    ctx.addr = mmvad.starting_vpn << 12;
+    size_t buffer_size = (mmvad.ending_vpn - mmvad.starting_vpn + 1) << 12;
+
+    if (!dump_memory_region(drakvuf, vmi, info, target->plugin, &ctx, buffer_size, "Usermode heuristic", nullptr, nullptr))
+    {
+        PRINT_DEBUG("[MEMDUMP] Failed to store memory dump due to an internal error\n");
+    }
+
+    return VMI_EVENT_RESPONSE_NONE;
+}
+
+/**
+ * Check if this thread is currently in process of loading a DLL.
+ * If so, return a pointer to the associated metadata.
+ */
+static user_dll_t* get_pending_dll(drakvuf_t drakvuf, drakvuf_trap_info* info, memdump* plugin)
+{
+    uint32_t thread_id;
+    if (!drakvuf_get_current_thread_id(drakvuf, info, &thread_id))
+        return nullptr;
+
+    auto vec_it = plugin->loaded_dlls.find(info->regs->cr3);
+
+    if (vec_it == plugin->loaded_dlls.end())
+        return nullptr;
+
+    for (auto& dll_meta : vec_it->second)
+    {
+        if (!dll_meta.is_hooked && dll_meta.thread_id == thread_id)
+            return &dll_meta;
+    }
+
+    return nullptr;
+}
+
+static bool populate_hook_targets(drakvuf_t drakvuf, memdump* plugin, const mmvad_info_t& mmvad, user_dll_t* dll_meta)
+{
+    vmi_instance_t vmi = drakvuf_lock_and_get_vmi(drakvuf);
+    unicode_string_t* dll_name = drakvuf_read_unicode_va(vmi, mmvad.file_name_ptr, 0);
```

Comment:
Leaks `dll_name`?

Reply:
This was fixed in some older revision but some refactor re-introduced a bug :( Fixed once again.